crates/bevy_ecs/src/system/query.rs

```diff
@@ -701,7 +702,10 @@ impl<'w, 's, D: QueryData, F: QueryFilter> Query<'w, 's, D, F> {
     ///
     /// [`iter`](Self::iter) for read-only query items.
     #[inline]
-    pub fn iter_mut(&mut self) -> QueryIter<'_, 's, D, F> {
+    pub fn iter_mut(&mut self) -> QueryIter<'_, 's, D, F>
+    where
+        D: IterQueryData,
```

Comment:
The PR looks good, but I don't fully understand `IterQueryData`.

The trait says 
```
A [`QueryData`] for which instances may be alive for different entities concurrently.
```

But when doing `iter_mut`, we never have two 2 items concurrently right? since we would iterate through them one by one. So even if we have a nested query which accesses multiple entities, there would still not be a collision?